// @openzeppelin/upgrades/lib/index.js
"use strict";
var __importDefault = (this && this.__importDefault) || function (mod) {
    return (mod && mod.__esModule) ? mod : { "default": mod };
};
Object.defineProperty(exports, "__esModule", { value: true });
// module information
const version = 'v' + require('../package.json').version;
exports.version = version;
// helpers
const encodeCall_1 = __importDefault(require("./helpers/encodeCall"));
exports.encodeCall = encodeCall_1.default;
const encodeCall_2 = require("./helpers/encodeCall");
exports.encodeParams = encodeCall_2.encodeParams;
const sleep_1 = __importDefault(require("./helpers/sleep"));
exports.sleep = sleep_1.default;
// utils
const ABIs_1 = __importDefault(require("./utils/ABIs"));
exports.ABI = ABIs_1.default;
const Semver_1 = __importDefault(require("./utils/Semver"));
exports.Semver = Semver_1.default;
const Logger_1 = require("./utils/Logger");
exports.Loggy = Logger_1.Loggy;
const Contracts_1 = __importDefault(require("./artifacts/Contracts"));
exports.Contracts = Contracts_1.default;
const Contract_1 = require("./artifacts/Contract");
exports.contractMethodsFromAst = Contract_1.contractMethodsFromAst;
exports.contractMethodsFromAbi = Contract_1.contractMethodsFromAbi;
exports.ContractMethodMutability = Contract_1.ContractMethodMutability;
exports.getConstructorInputs = Contract_1.getConstructorInputs;
const ZWeb3_1 = __importDefault(require("./artifacts/ZWeb3"));
exports.ZWeb3 = ZWeb3_1.default;
const Bytecode_1 = require("./utils/Bytecode");
exports.bodyCode = Bytecode_1.bodyCode;
exports.constructorCode = Bytecode_1.constructorCode;
exports.bytecodeDigest = Bytecode_1.bytecodeDigest;
exports.replaceSolidityLibAddress = Bytecode_1.replaceSolidityLibAddress;
exports.isSolidityLib = Bytecode_1.isSolidityLib;
exports.getSolidityLibNames = Bytecode_1.getSolidityLibNames;
const Transactions_1 = __importDefault(require("./utils/Transactions"));
exports.Transactions = Transactions_1.default;
const Solidity_1 = require("./utils/Solidity");
exports.flattenSourceCode = Solidity_1.flattenSourceCode;
const Semver_2 = require("./utils/Semver");
exports.semanticVersionEqual = Semver_2.semanticVersionEqual;
exports.toSemanticVersion = Semver_2.toSemanticVersion;
exports.semanticVersionToString = Semver_2.semanticVersionToString;
const Migrator_1 = __importDefault(require("./utils/Migrator"));
exports.AppProxyMigrator = Migrator_1.default;
// validations
const Storage_1 = require("./validations/Storage");
exports.getStorageLayout = Storage_1.getStorageLayout;
exports.getStructsOrEnums = Storage_1.getStructsOrEnums;
const BuildArtifacts_1 = require("./artifacts/BuildArtifacts");
exports.getBuildArtifacts = BuildArtifacts_1.getBuildArtifacts;
exports.BuildArtifacts = BuildArtifacts_1.BuildArtifacts;
const Layout_1 = require("./validations/Layout");
exports.compareStorageLayouts = Layout_1.compareStorageLayouts;
const validations_1 = require("./validations");
exports.validate = validations_1.validate;
exports.newValidationErrors = validations_1.newValidationErrors;
exports.validationPasses = validations_1.validationPasses;
// test behaviors
const test_1 = require("./test");
exports.behaviors = test_1.behaviors;
exports.helpers = test_1.helpers;
const assertions = test_1.helpers.assertions;
exports.assertions = assertions;
const assertRevert = test_1.helpers.assertRevert;
exports.assertRevert = assertRevert;
const assertEvent = test_1.helpers.assertEvent;
exports.assertEvent = assertEvent;
// model objects
const Proxy_1 = __importDefault(require("./proxy/Proxy"));
exports.Proxy = Proxy_1.default;
const MinimalProxy_1 = __importDefault(require("./proxy/MinimalProxy"));
exports.MinimalProxy = MinimalProxy_1.default;
const ProxyAdmin_1 = __importDefault(require("./proxy/ProxyAdmin"));
exports.ProxyAdmin = ProxyAdmin_1.default;
const ProxyFactory_1 = __importDefault(require("./proxy/ProxyFactory"));
exports.ProxyFactory = ProxyFactory_1.default;
const App_1 = __importDefault(require("./application/App"));
exports.App = App_1.default;
const Package_1 = __importDefault(require("./application/Package"));
exports.Package = Package_1.default;
const ImplementationDirectory_1 = __importDefault(require("./application/ImplementationDirectory"));
exports.ImplementationDirectory = ImplementationDirectory_1.default;
const BasePackageProject_1 = __importDefault(require("./project/BasePackageProject"));
exports.BasePackageProject = BasePackageProject_1.default;
const PackageProject_1 = __importDefault(require("./project/PackageProject"));
exports.PackageProject = PackageProject_1.default;
const AppProject_1 = __importDefault(require("./project/AppProject"));
exports.AppProject = AppProject_1.default;
const SimpleProject_1 = __importDefault(require("./project/SimpleProject"));
exports.SimpleProject = SimpleProject_1.default;
const ProxyAdminProject_1 = __importDefault(require("./project/ProxyAdminProject"));
exports.ProxyAdminProject = ProxyAdminProject_1.default;
