// @nomiclabs/buidler/builtin-tasks/task-names.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
exports.TASK_CHECK = "check";
exports.TASK_CLEAN = "clean";
exports.TASK_COMPILE = "compile";
exports.TASK_COMPILE_GET_SOURCE_PATHS = "compile:get-source-paths";
exports.TASK_COMPILE_GET_RESOLVED_SOURCES = "compile:get-resolved-sources";
exports.TASK_COMPILE_GET_DEPENDENCY_GRAPH = "compile:get-dependency-graph";
exports.TASK_COMPILE_GET_COMPILER_INPUT = "compile:get-compiler-input";
exports.TASK_COMPILE_RUN_COMPILER = "compile:run-compiler";
exports.TASK_COMPILE_COMPILE = "compile:compile";
exports.TASK_COMPILE_CHECK_CACHE = "compile:cache";
exports.TASK_BUILD_ARTIFACTS = "compile:build-artifacts";
exports.TASK_CONSOLE = "console";
exports.TASK_FLATTEN = "flatten";
exports.TASK_FLATTEN_GET_FLATTENED_SOURCE = "flatten:get-flattened-sources";
exports.TASK_HELP = "help";
exports.TASK_RUN = "run";
exports.TASK_NODE = "node";
exports.TASK_TEST = "test";
exports.TASK_TEST_RUN_MOCHA_TESTS = "test:run-mocha-tests";
exports.TASK_TEST_GET_TEST_FILES = "test:get-test-files";
exports.TASK_TEST_SETUP_TEST_ENVIRONMENT = "test:setup-test-environment";
