// @nomiclabs/buidler/internal/solidity/compiler/compiler-input.js
"use strict";
Object.defineProperty(exports, "__esModule", { value: true });
function getInputFromDependencyGraph(graph, optimizerConfig, evmVersion) {
    const sources = {};
    for (const file of graph.getResolvedFiles()) {
        sources[file.globalName] = {
            content: file.content,
        };
    }
    const input = {
        language: "Solidity",
        sources,
        settings: {
            metadata: {
                useLiteralContent: true,
            },
            optimizer: optimizerConfig,
            outputSelection: {
                "*": {
                    "*": [
                        "abi",
                        "evm.bytecode",
                        "evm.deployedBytecode",
                        "evm.methodIdentifiers",
                    ],
                    "": ["id", "ast"],
                },
            },
        },
    };
    if (evmVersion !== undefined) {
        input.settings.evmVersion = evmVersion;
    }
    return input;
}
exports.getInputFromDependencyGraph = getInputFromDependencyGraph;
